### ISSUE TITLE
live-update diagnostics

### ISSUE THREAD

**Issue by dakom** [rust-analyzer/rust-analyzer#3107]
Is it possible to get diagnostics when text is edited, like with a debounce/delay so its not overwhelming?

Right now I have to save in order to see them - which is more of a problem than it may seem since old errors lurk around and pollute the screen

**Comment by bjorn3** [comment]
These diagnostics come from running `cargo check`. This needs the current code to be saved. Rls solves this problem by hooking into nightly only compiler api's to present a virtual fs containing the current unsaved code.

**Comment by dakom** [comment]
thanks for the insight! is this something rust-analyzer could do too?

**Comment by bjorn3** [comment]
Currently rust-analyzer is using stable. These api's however require nightly, so using them is not possible. Using for example LD_PRELOAD it would be possible to change the open syscall, but this is a hack and I don't know if windows has something equivalent.

**Comment by Immugio** [comment]
@bjorn3 Is there any plan to make the analysis work while typing?

**Comment by resolritter** [comment]
> These diagnostics come from running `cargo check`. This needs the current code to be saved. Rls solves this problem by hooking into nightly only compiler api's to present a virtual fs containing the current unsaved code.

Related to this, on RLS 

https://github.com/rust-lang/rls/blob/67bce0bdcf2db1d3c05bb1a3d87df9e66eaec7db/src/build/rustc.rs#L132

Notice how it  using this `run_compiler` API from `rustc_driver`.

https://github.com/rust-lang/rustc-dev-guide/blob/master/src/rustc-driver.md
https://github.com/rust-lang/rustc-dev-guide/blob/master/src/rustc-driver-getting-diagnostics.md

Due to my unfamiliarity with Rust, I couldn't find out whether `run_compiler` is still nightly-only (it seems to be). I also could not figure out when it'd be stabilized from the issue tracker. @bjorn3 since you have more context, if you happen to have a `rustc` issue in mind which is blocking this, it'd be good to link it here for the subscribed people.

**Comment by bjorn3** [comment]
`run_compiler` is a permanently unstable function. We want to be able to change it in the future. In fact I changed it a few days ago to add another parameter.

**Comment by resolritter** [comment]
> `run_compiler` is a permanently unstable function. We want to be able to change it in the future. In fact I changed it a few days ago to add another parameter.

@bjorn3  Thanks for the clarification. My attempt was for understanding what is blocking progress on this issue on the `rustc` side; knowing that makes contributing easier and is also useful for this issue's subscribers. For instance, quoting https://github.com/rust-analyzer/rust-analyzer/issues/3107#issuecomment-584623437

> These api's however require nightly

Is there some roadmap and/or tracking issue for having them on stable? (Or maybe they already are?)

**Comment by bjorn3** [comment]
No, there is no roadmap, nor a tracking issue. These are implementation details of the rust compiler. It is highly likely that the api's necessary for this will look fundamentally different in a few years. Either as a result of the ongoing libraryfication of rustc (to make it easier to share code with other tools like rust-analyzer) or as a result of the wish to push back incremental compilation more towards the macro expansion/parsing part of the compilation.

**Comment by Kimundi** [comment]
@bjorn3 Would adding a new high-level CLI flag to `rustc` that allows overriding filesystem paths read by the compiler be something thats high-level enough to be stabilizable?

For example:

```
> rustc main.rs --override_paths=config.json
```

Source directory:
```
main.rs
submodule.rs
```

config.json
```json
{
    "overrides": [
        { "path": "submodule.rs", "override": "/some/tmp/ra/dir/submodule.rs" }
    ]
}
```

**Comment by MikailBag** [comment]
I think such a remapping can be already achieved using e.g. bind mounts (at least on Linux).

**Comment by lukechu10** [comment]
Is there any reason preventing rust-analyzer from using nightly compiler APIs?
I think there are two ways forward from here:
- Using the nightly only `run_compiler` from `rustc_driver`.
- Exposing a higher-level API from rustc that is stable.

**Comment by flodiebold** [comment]
> Is there any reason preventing rust-analyzer from using nightly compiler APIs?

There are reasons, yes -- we prefer rust-analyzer to build on stable with normal `cargo build` to make contribution as easy as possible. Also, as long as our release cycle isn't aligned with rustc's, we can't really be bound to a specific rustc version.

These problems might be avoidable by implementing a separate custom driver for rustc, but the other reason why no-one is spending effort on this is that it's not the final architecture we intend. The whole reason for rust-analyzer's existence is that the RLS approach of just running the compiler on every change is a dead-end. The current 'cargo watch' / flycheck integration is a stopgap to provide diagnostics that rust-analyzer can't provide natively, but our focus is on librarification of rustc and providing more and more diagnostics natively, ideally by sharing library code with rustc. We're not far off from showing type errors natively (without saving), for example.

(Just to be clear, that's not saying that any particular pull requests would be rejected out of hand.)

**Comment by flodiebold** [comment]
This already exists as an experimental diagnostic; you can enable it by setting `rust-analyzer.diagnostics.experimental.enable` to true. You might get some false positive errors though, which is why it's not enabled by default yet.

**Comment by emi2k01** [comment]
@flodiebold, awesome! thanks :)

**Comment by tats-u** [comment]
This issue should be pinned because it gains the most numbers of 👍️.
Almost all of language servers of other (major) languages can but this Rust analyzers can't.

**Comment by ChayimFriedman2** [comment]
@tats-u This issue does not get updated when we add more diagnostics, so I see no reason to pin it.

Achieving parity with rustc's diagnostics is a very big project that's going to take a lot of time. We are incrementally adding more and more diagnostics. Next-gen trait solver (on which work was just started) may finally allow us to enable typecheck diagnostics by default, and borrow check diagnostics will likely require more work or switch to Polonius before enabling by default. These are the bigger ones, but there are many small ones.
